src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,553 @@
+  if (my_r != r) {
+    UMPIRE_ERROR(
+        runtime_error,
+        fmt::format("Called deallocate with different resource than what is returned by getResource. Called with {},",
+                    "but getResource returned: {}", camp::resources::to_string(r), camp::resources::to_string(my_r)));
+  }
+
+  if (m_is_coalescing == false) {
+    chunk->m_event = r.get_event();
+  }
+
+  m_used_map.erase(ptr);
+  m_current_bytes -= chunk->size;
+
+  // Call deallocate logic only for a non-pending chunk
+  if (chunk->m_event.check()) {
+    do_deallocate(chunk, ptr);
+  } else {
+    // Chunk is now pending, add to list
+    m_pending_map.push_back(chunk);
+  }
+
+  std::size_t suggested_size{m_should_coalesce(*this)};
+  if (0 != suggested_size) {
+    UMPIRE_LOG(Debug, "coalesce heuristic true, performing coalesce.");
+    do_coalesce(suggested_size);
+  }
+}
+
+void ResourceAwarePool::release()
+{
+  UMPIRE_LOG(Debug, "() " << m_free_map.size() << " chunks in free map, m_is_destructing set to " << m_is_destructing);
+
+#if defined(UMPIRE_ENABLE_BACKTRACE)
+  std::size_t prev_size{m_actual_bytes};
+#endif
+
+  // TODO:
+  // This will check all chunks in m_pending_map and erase the entry if event is complete
+  for (auto it = m_pending_map.begin(); it != m_pending_map.end();) {
+    auto chunk = (*it);
+    if (chunk != nullptr && chunk->free == false && chunk->m_event.check()) {
+      m_free_map.insert(std::make_pair(chunk->size, chunk));
+      chunk->free = true;
+      m_pending_map.erase(it);
+    } else {
+      it++;
+    }
+  }
+
+  for (auto pair = m_free_map.begin(); pair != m_free_map.end();) {
+    auto chunk = (*pair).second;
+    UMPIRE_LOG(Debug, "Found chunk @ " << chunk->data);
+    if ((chunk->size == chunk->chunk_size) && chunk->free) {
+      UMPIRE_LOG(Debug, "Releasing chunk " << chunk->data);
+
+      m_actual_bytes -= chunk->chunk_size;
+      m_releasable_bytes -= chunk->chunk_size;
+      m_releasable_blocks--;
+      m_total_blocks--;
+
+      try {
+        aligned_deallocate(chunk->data);
+      } catch (...) {
+        if (m_is_destructing) {
+          //
+          // Ignore error in case the underlying vendor API has already shutdown
+          //
+          UMPIRE_LOG(Error, "Pool is destructing, runtime_error Ignored");
+        } else {
+          throw;
+        }
+      }
+
+      chunk->~Chunk(); // manually call destructor
+      m_chunk_pool.deallocate(chunk);
+      pair = m_free_map.erase(pair);
+    } else {
+      ++pair;
+    }
+  }
+
+#if defined(UMPIRE_ENABLE_BACKTRACE)
+  if (prev_size > m_actual_bytes) {
+    umpire::util::backtrace bt;
+    umpire::util::backtracer<>::get_backtrace(bt);
+    UMPIRE_LOG(Info, "actual_size:" << m_actual_bytes << " (prev: " << prev_size << ") "
+                                    << umpire::util::backtracer<>::print(bt));
+  }
+#endif
+}
+
+std::size_t ResourceAwarePool::getReleasableBlocks() const noexcept
+{
+  return m_releasable_blocks;
```

Comment:
Are blocks with only pending allocations considered releasable?

Reply:
That is a great question. I would say no, not until they have been actually deallocated. @davidbeckingsale what do you think?